File: sprint 2/jira.js
async function getJiraIssues() {
  // This function is designed to fetch issues from a Jira URL provided by the user using a POST request.
  // It takes user input for the Jira URL, API token, and user email.
  // It then sends a POST request to the local server to fetch issues from the Jira API.
  // If the request is successful, it displays the issues in a table format.
  // If there are no issues found, it displays a message indicating that no issues were found.
  // If there is an error, it displays an error message.
  const jiraUrl = document.getElementById('jira-url').value;
  const apiToken = document.getElementById('api-token').value;
  const userEmail = document.getElementById('user-email').value;

  const baseUrl = jiraUrl.replace(/\/jira\/software\/projects\/.*$/, '/rest/api/3/search');
  const errorContainer = document.getElementById('error-container');
  const issuesList = document.getElementById('issues-list');
  const issuesTableHeader = document.getElementById('issues-table-header');
  
  errorContainer.innerHTML = ''; // Clear previous error messages
  issuesList.innerHTML = ''; // Clear previous issues
  issuesTableHeader.innerHTML = ''; // Clear previous table header

  if (!baseUrl.includes('/rest/api/3/search')) {
    errorContainer.innerHTML = '<div class="error-message">Invalid URL</div>';
    return;
  }

  try {
    const response = await fetch('http://127.0.0.1:5000/api/issues', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ baseUrl, apiToken, userEmail }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(errorText);
    }

    const issues = await response.json();

    if (issues.length === 0) {
      issuesList.innerHTML = '<h2>No issues found for this project</h2>';
      return;
    }

    issuesTableHeader.innerHTML = `
      <tr>
        <th>Issue Key</th>
        <th>Summary</th>
        <th>Assignee</th>
        <th>Reporter</th>
        <th>Status</th>
        <th>Created</th>
        <th>Resolved</th>
        <th>Duration</th>
      </tr>
    `;

    issuesList.innerHTML = issues.map(issue => {
      let statusClass = '';
      switch (issue.status.toLowerCase()) {
        case 'todo':
          statusClass = 'status-todo';
          break;
        case 'in progress':
          statusClass = 'status-in-progress';
          break;
        case 'done':
          statusClass = 'status-done';
          break;
        case 'dropped':
          statusClass = 'status-dropped';
          break;
        default:
          statusClass = '';
      }

      const createdDate = new Date(issue.created);
      const resolvedDate = issue.resolved !== "Unresolved" ? new Date(issue.resolved) : null;
      let duration = "Unresolved";
      if (resolvedDate) {
        const diffTime = Math.abs(resolvedDate - createdDate);
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        const diffHours = Math.ceil(diffTime / (1000 * 60 * 60));
        if (diffDays > 1) {
          duration = `${diffDays} days`;
        } else if (diffDays === 1) {
          duration = "1 day";
        } else {
          duration = `${diffHours} hours`;
        }
      }

      return `
        <tr>
          <td>${issue.key}</td>
          <td class="summary">${issue.summary}</td>
          <td>${issue.assignee}</td>
          <td>${issue.reporter}</td>
          <td><span class="issue-status ${statusClass}">${issue.status}</span></td>
          <td>${createdDate.toLocaleDateString()}</td>
          <td>${resolvedDate ? resolvedDate.toLocaleDateString() : "Unresolved"}</td>
          <td>${duration}</td>
        </tr>
      `;
    }).join('');
  } catch (error) {
    errorContainer.innerHTML = `<div class="error-message">${error.message}</div>`;
  }
}

File: sprint 2/jira.py
import requests, json
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)  # Enable CORS for all routes

@app.route('/api/issues', methods=['POST'])
def get_issues():
    # Get the data from the request input
    data = request.json
    base_url = data.get('baseUrl')
    user = data.get('userEmail')
    pwd = data.get('apiToken')

    if not base_url or not user or not pwd:
        return "Invalid Token or Email", 400

    try:
        response = requests.get(base_url, headers={
            "Accept": "application/json",
            "Content": "application/json"
        }, auth=(user, pwd))

        if response.status_code == 401:
            return "This token is not authorized for this Jira URL", 401
        elif response.status_code == 403:
            return "The email is not connected or authorized to the said Jira URL", 403
        elif response.status_code != 200:
            return "Failed to fetch issues", response.status_code

        issues = response.json()["issues"]
        # Extract the necessary information from the issues and return it
        issues_list = [
            {
                "key": issue["key"],
                "summary": issue["fields"]["summary"],
                "status": issue["fields"]["status"]["name"],
                "reporter": issue["fields"]["reporter"]["displayName"] if issue["fields"]["reporter"] else "Unknown",
                "created": issue["fields"]["created"],
                "resolved": issue["fields"]["resolutiondate"] if issue["fields"]["resolutiondate"] else "Unresolved",
                "assignee": issue["fields"]["assignee"]["displayName"] if issue["fields"]["assignee"] else "Unassigned"
            }
            for issue in issues
        ]
        return jsonify(issues_list)
    except requests.exceptions.RequestException as e:
        return str(e), 500

if __name__ == '__main__':
    app.run(debug=True)